File: python/p12/p12.py
# http://projecteuler.net/problem=12

import itertools
import math


def triangle_number(n):
    """ Returns the nth triangle number """
    return n * (n + 1) / 2

def divisors(n):
    """ Find the divisors (and their n / divisor complement) for the
        numbers up to sqrt(n) + 1 """
    d = []
    for i in range(1, int(math.sqrt(n) + 1)):
        if n % i == 0:
            d.append(i)
            d.append(n / i)
    return set(d)


# Find the first triangle number with more than 500 divisors

for i in itertools.count(1):
    t = triangle_number(i)
    if len(divisors(t)) > 500:
        print t
        break

File: csharp/P11_to_20/P12.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ProjectEuler.Helpers;

namespace ProjectEuler.P11_to_20
{
    /// <summary>
    /// https://projecteuler.net/problem=12
    /// </summary>
    internal class P12
    {
        /// <summary>
        /// Find the first triangle number to have over 500 divisors.
        /// </summary>
        public long Solve()
        {
            var counter = 1;
            var triangleNumber = 1L;

            while (true)
            {
                var divisors = PrimeHelpers.Divisors(triangleNumber);
                if (divisors.Length > 500)
                {
                    return triangleNumber;
                }

                counter++;
                triangleNumber += counter;
            }
        }
    }
}
